### ISSUE TITLE
Revise community roles

### ISSUE THREAD

**Issue by robzienert** [spinnaker/governance#157]
I'll be submitting a PR (or series of PRs) to update the community roles of Spinnaker's governance. Our community roles are not as well-defined as they should be, and it is often not clear what the responsibilities and privileges are of any particular role. The governance repo was originally drawn from the [Kubernetes community](https://github.com/kubernetes/community), and this change will also be taking inspiration (in some cases, direct) from changes they have made since.

I plan to do the following things, having discussed previously with the Technical Oversight Committee and other members of the community:

1. Introduce a new *Member* community role, which will be the official, low-bar entry point into the Spinnaker contributors community.
1. Clearly define the ladder of community roles; and the paths that one can take to reach a particular role.
1. Revise the requirements for the Reviewer and Approver roles to be more qualitative.
1. Introduce new, clearly defined responsibilities and privileges for each community role.
1. Introduce a direct path from Member role to Approver role under limited conditions.
1. Update the TOC and Steering Committee governance language to reflect the format and structure of other community roles.
1. Alter the scope of both Reviewer and Approver to be per-project, moving the source of truth to `CODEOWNERS` files in each repository.
1. Introduce new mailing lists to help funnel announcements and collaboration through a platform more accessible than Slack.
1. Define a clear process for revocation (and restoration) of community roles.

There has been discussion of removing the Reviewer role entirely, however I am making an effort to re-define it in such a way that it has clearer value than it does today.

**Comment by ajordens** [comment]
I like it ... I've been thinking about per-project for a couple of the non-core services so this makes sense. Perhaps it's actually worth drawing a distinction between core/non-core (the latter possibly including kayenta, rosco, spin, etc.).

We should also define a process that regularly reviews approvers and revokes access if they are no longer actively involved. Perhaps with wording that there's a straight forward path to earning them back.

On #7, we talked offline but I believe we mean `CODEOWNERS`.

**Comment by markyjackson-taulia** [comment]
The reviewer role is a good Segway to approver and is more of a triage for the approver.
It also helps to clearly define who can lgtm a pr and who can approve.
Finally, reviewer status helps prepare members on how to properly do a code review so it also serves as a mentoring aspect.

**Comment by ezimanyi** [comment]
I just read through the Kubernetes requirements in detail, and some of the things that I like from their process are:
* The level of "member" allows contributions of multiple types (ex: opening issues, commenting on other issues, etc.).
  * There are a number of members of the Spinnaker community who don't often submit code or review PRs but are extremely knowledgable/helpful in Slack and Github issues, and it would be great be able to recognize this.
* There is a time-based process that requires contribution at one particular level before approval for the next level.
  * I kind of like the idea of having people demonstrate their participation at one level for some amount of time before moving to the next.
* You can be [be nominated by a robot](https://github.com/kubernetes/community/blob/master/community-membership.md#requirements-1) 🤖 
  * I have a lot of questions

**Comment by robzienert** [comment]
Adding a little more context on #7: I'm trying to follow a little bit more in the path of Kubernetes community here where they have an `OWNERS` file that has a list of reviewers and approvers. `CODEOWNERS` support in Github isn't as flexible as what I'd like: It basically just allows us to auto-request reviewership, and optionally require approval from an owner. What I'd like to do instead is write a Github Action that will read a `MEMBERS` file in the root of a project that looks kinda like so:

```yaml
some/path/foo/:
  reviewers:
  - robzienert
  approvers:
  - ajordens
  notify:
  - @spinnaker/sig-ci 
```

The paths would follow the same pattern matching that `CODEOWNERS` supports. `notify` would be there to loop more people in that may be interested in an area, but who are not explicitly filling the reviewer role for that particular area of the project.

I'm happy to pull this out into a separate issue if we wanna dig more at it.

**Comment by robzienert** [comment]
> I kind of like the idea of having people demonstrate their participation at one level for some amount of time before moving to the next.

@ezimanyi Agreed. In my current draft, I have time-based process entering into Approver, but not into Reviewer (like Kubernetes has). I also have an inactivity window of 12 months vs their 18. I felt like adding _too_ much time-based process isn't well suited to our community size, whereas Kubernetes probably has an issue of too many people trying to get reviewer/approvership. I'm open to being swayed in any direction here.

> You can be be nominated by a robot 🤖

Naturally, I left adopting this out, but I like the idea of it. We have plenty of stats via [DevStats](https://spinnaker.devstats.cd.foundation/d/8/dashboards?orgId=1&refresh=15m), it would be neat to automatically nominate people who have qualified for a particular bump for N+ time. I bet there's a few people out there that don't know they're eligible now for a role bump that this could help with. Sounds like a gardening day kinda thing, though... or not even that.

**Comment by markyjackson-taulia** [comment]
If I can help in any way, please do let me know.
I helped a lot on the Kubernetes and Jenkins side in terms of the technical items here.

**Comment by ajordens** [comment]
What do we see as the order of operations? 

The potential _monorepo_ refactor has some impact here and it might not be worth pushing `CODEOWNER` changes everywhere only to have consolidation down the road.

I see big value in role formalization. Perhaps we can knock that out first, seed a `MEMBERS` file and write a bit of automation to keep GitHub groups up-to-date. Once the _monorepo_ work is complete, we can update to generate appropriate `CODEOWNERS`.

**Comment by robzienert** [comment]
Yeah, I think I'm going to push out the substance of these language updates, sans any mention of `CODEOWNERS`, and then work on figuring that out later.

I think each repo could have a `MEMBERS` file, seeded as we desire (each repo gets a GitHub team for approver (e.g. `clouddriver-approvers`). If we collapse into a monorepo, I think we'd need to switch to using bot commands to merge. Even in a monorepo world, `CODEOWNERS` won't really help us much for managing reviewer/approver distinctions.

**Comment by robzienert** [comment]
I don't think it's worth changing anything regarding how we manage membership (as far as files, teams, etc go) until we have a clear direction on the monorepo proposal.

**Comment by ezimanyi** [comment]
Agreed on not making any changes until we have decided whether (and how) to move to a monorepo.  But even in that case I think we could still use `CODEOWNERS`, as I believe the syntax allows scoping owners to specific directories.

That, combined with a setting like this on the repo would probably achieve what we want:
![image](https://user-images.githubusercontent.com/1391982/88827507-08bc1700-d198-11ea-99ed-52c7a1077cbc.png)

**Comment by robzienert** [comment]
@ezimanyi This works for passing review, but not for granting write permissions to the repo, or in a monorepo case, write permissions to subprojects. I think Approvers might wind up just getting `triage` permissions, and we proxy writes through a bot (`/merge` or something)?

**Comment by ezimanyi** [comment]
Oh I see, yeah I guess we could either:
* Give everyone who is an approver anywhere write access but rely on the above setting to prevent them from merging things that are not reviewed by the appropriate code owner.
  * If I'm understanding correctly, with that setting, people with write access will still be unable to merge PRs that don't have the appropriate sign-offs and will also be prevented from directly pushing to `master`.
* Give nobody write access and rely on a bot to actually merge, as you suggest
  * There's probably a way to get `mergify` to fit our use case so we don't have to write our own.

**Comment by ajordens** [comment]
One potential annoyance could be the sheer size of `CODEOWNERS` and number of folks that would be tagged on PRs, it may not be all that mangeable. 

The bot merge + github action pr setup may be the way to go. Poked around at GitHub action marketplace and it looked like there were some that could do random PR assignment out of a groups of people. Perhaps that's a workable pattern.

**Comment by dnilasor** [comment]
CC @prophen - Folks this is our new OSS Community Marketing Associate, Nikema. She is passionate about this and other CX issues around creating a welcoming and inclusive project space for contributors, so I'm asking her to catch up on this.

**Comment by ajordens** [comment]
For now, I'm going to close this issue as the effort required to make this happen seamlessly is more than we time than we currently have.

Context will remain should the community wish to revisit this in the future.
